nuxeo-core/nuxeo-core-schema/src/main/java/org/nuxeo/ecm/core/schema/SchemaManagerImpl.java

```diff
@@ -280,13 +283,57 @@ protected void recompute() {
         recomputeFacets(); // depend on schemas
         recomputeDocumentTypes(); // depend on schemas and facets
         recomputeProxies(); // depend on schemas
+        recomputeDeprecatedFields(); // depend on schemas
         fields.clear(); // re-filled lazily
     }
 
     /*
      * ===== Configuration =====
      */
 
+    protected void recomputeDeprecatedFields() {
+        int step = 2;
+        if (step == 1) {
+            // First step, declare them manually
+            // Removed property
+            Schema commonSchema = schemas.get("common");
+            commonSchema.addField("size", RemovedType.INSTANCE, null, Field.DEPRECATED, null);
+            // Removed property with fallback
+            Schema fileSchema = schemas.get("file");
+            fileSchema.addField("filename", RemovedType.INSTANCE, null, Field.DEPRECATED, null);
+            // Deprecated property with fallback
+
+        } else if (step == 2) {
+            // Second step, declare them as a list
+            List<String> properties = Arrays.asList("common:size", "file:filename", "deprecated:scalar2scalar", "deprecated:scalar2complex", "deprecated:complex");
+            Map<String, List<String>> propertiesBySchema = properties.stream().collect(
+                    Collectors.groupingBy(p -> p.substring(0, p.indexOf(':')),
+                            Collectors.mapping(p -> p.substring(p.indexOf(':') + 1), Collectors.toList())));
+            for (Entry<String, List<String>> entry : propertiesBySchema.entrySet()) {
+                // First try to locate the deprecated property
+                Schema schema = schemas.get(entry.getKey());
+                // TODO do we handle schema deprecation ? or just throw an exception
+                for (String fieldName : entry.getValue()) {
+                    Field field = schema.getField(fieldName);
+                    if (field == null) {
+                        field = schema.addField(fieldName, RemovedType.INSTANCE, null, Field.DEPRECATED, null);
+                    } else {
+                        field.setDeprecated(true);
+                    }
+                    if ("scalar2scalar".equals(fieldName)) {
+                        field.setFallbackXpath("scalarfallback");
+                    } else if ("scalar2complex".equals(fieldName)) {
+                        field.setFallbackXpath("complexfallback/scalar");
+                    } else if ("complex".equals(fieldName)) {
+                        field.setFallbackXpath("complexfallback");
+                    } else if ("filename".equals(fieldName)) {
+                        field.setFallbackXpath("content/name");
+                    }
```

Comment:
Here we can't use a real xpath (naming to change), because we can't fallback on a property which doesn't belong to the same schema.
I assume that for this first version we won't handle schema deprecation (not the scope) and not handle fallback in a different schema.